pkg/gpu/stream.go

```diff
@@ -0,0 +1,178 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package gpu
+
+import (
+	"math"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/gpu/model"
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	gpuebpf "github.com/DataDog/datadog-agent/pkg/gpu/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// StreamHandler is responsible for receiving events from a single CUDA stream and generating
+// stats from them.
+type StreamHandler struct {
+	kernelLaunches []gpuebpf.CudaKernelLaunch
+	memAllocEvents map[uint64]gpuebpf.CudaMemEvent
+	kernelSpans    []*model.KernelSpan
+	allocations    []*model.MemoryAllocation
+	processEnded   bool // A marker to indicate that the process has ended, and this handler should be flushed
+}
+
+func newStreamHandler() *StreamHandler {
+	return &StreamHandler{
+		memAllocEvents: make(map[uint64]gpuebpf.CudaMemEvent),
+	}
+}
+
+func (sh *StreamHandler) handleKernelLaunch(event *gpuebpf.CudaKernelLaunch) {
+	// Copy events, as the memory can be overwritten in the ring buffer after the function returns
+	sh.kernelLaunches = append(sh.kernelLaunches, *event)
+}
+
+func (sh *StreamHandler) handleMemEvent(event *gpuebpf.CudaMemEvent) {
+	if event.Type == gpuebpf.CudaMemAlloc {
+		sh.memAllocEvents[event.Addr] = *event
+		return
+	}
+
+	alloc, ok := sh.memAllocEvents[event.Addr]
```

Comment:
if that is not CudaMemAlloc it must be CudaFree? if that's the case, please add an explicit comment.

P.S: I think we want to hook more memory management functions of CUDA

Reply:
Added. I will look into more functions in a followup PR.